feat: internal provider state, new client events (#241)
open-feature/spec: specification/sections/02-providers.md

}
```

#### Condition 2.4.2

> The provider defines an `initialize` function.

##### Conditional Requirement 2.4.2.1

> > If the provider's `initialize` function fails to ready the provider to evaluate flags, it **MUST** abnormally terminate.

Reviewer: I'm not sure what this section means. Could you please rephrase?

}
```

#### Condition 2.4.2

> The provider defines an `initialize` function.

##### Conditional Requirement 2.4.2.1

> If the provider's `initialize` function fails to render the provider ready to evaluate flags, it **SHOULD** abnormally terminate.

If a provider is unable to start up correctly, it should indicate abnormal execution by throwing an exception, returning an error, or otherwise indicating so by means idiomatic to the implementation language.

> If the error is irrecoverable (perhaps due to bad credentials or invalid configuration) you can use the `PROVIDER_FATAL` error code.

Reviewer: ```suggestion
If the error is irrecoverable (perhaps due to bad credentials or invalid configuration) the `PROVIDER_FATAL` error code should be used.
```

Author: nice catch.

> If the error is irrecoverable (perhaps due to bad credentials or invalid configuration) you can use the `PROVIDER_STATE_FATAL` error code.

Reviewer 2: I assume the implementation here will come in a form of an exception type (ex:-`ProviederStateFatalExpection.java`) ?

Author: > I assume the implementation here will come in the form of an exception/error type (ex:-ProviederStateFatalExpection.java) ?

Yes.

> Besides, does this mean SDK should differentiate normal error vs PROVIDER_STATE_FATAL and attempt retries. I think SDK shouldn't but this could happen at provider eventually making it to ready status isn't it ?

The provider handles it's own business concerning reconnects and such internal. If it emits a `FATAL` error, it would be logically inconsistent for it to attempt to fix that state. If it throws FATAL, it should give up, and the SDK would make that error code known to the application author/integrator as per the new section `1.7.8`.

Reviewer 2: Yes, concern is resolved with the updated requirement.
